File: wagerproof-ios-native/Wagerproof/Features/Props/Detail/RecentPropBarChart.swift
import SwiftUI
import Charts
import WagerproofModels
import WagerproofDesign

/// Recent-form bar chart for a player prop. Ports
/// `wagerproof-mobile/components/mlb/player-props/RecentPropBarChart.tsx`
/// from react-native-svg to Apple's `Charts` framework.
///
/// The behaviour the user asked for lives here: as the selected line changes
/// (via the slider/chips upstream), the dashed threshold line slides and the
/// bars recolor green/red in real time. We drive that by recomputing
/// `bars` + `line` upstream and animating the value change so SwiftUI
/// crossfades the colours and tweens the rule position.
///
/// Scaling mirrors RN exactly: `maxVal = max(line*1.5, maxBarValue, line+1, 1)`
/// keeps the threshold at a stable visual height as the line moves.
struct RecentPropBarChart: View {
    let bars: [MLBPropChartBar]
    let line: Double

    private var maxVal: Double {
        let vals = bars.map(\.value)
        return max(line * 1.5, (vals.max() ?? 0), line + 1, 1)
    }

    var body: some View {
        if bars.isEmpty {
            Text("No recent games")
                .font(.system(size: 13))
                .italic()
                .foregroundStyle(Color.appTextMuted)
                .frame(maxWidth: .infinity, alignment: .leading)
        } else {
            chart
                .frame(height: 168)
        }
    }

    private var chart: some View {
        Chart {
            ForEach(bars) { bar in
                BarMark(
                    x: .value("Game", String(bar.id)),
                    y: .value("Value", bar.value),
                    width: .ratio(0.62)
                )
                .cornerRadius(2)
                .foregroundStyle(bar.cleared ? Color.appPrimary : Color.appLoss.opacity(0.7))
                .annotation(position: .top, spacing: 2) {
                    Text(MLBPlayerProps.formatBarValue(bar.value))
                        .font(.system(size: 9, weight: .bold))
                        .foregroundStyle(bar.cleared ? Color.appPrimary : Color.appLoss)
                }
            }

            // Dashed threshold for the currently-selected line.
            RuleMark(y: .value("Line", line))
                .lineStyle(StrokeStyle(lineWidth: 1.2, dash: [4, 3]))
                .foregroundStyle(Color.appPrimary.opacity(0.85))
                .annotation(position: .top, alignment: .trailing, spacing: 1) {
                    Text("Line \(MLBPlayerProps.formatBarValue(line))")
                        .font(.system(size: 9, weight: .bold))
                        .foregroundStyle(Color.appPrimary)
                }
        }
        // Fixed domain (oldest→newest) so bars never reorder as the line moves.
        .chartXScale(domain: bars.map { String($0.id) })
        .chartYScale(domain: 0...maxVal)
        .chartYAxis(.hidden)
        .chartLegend(.hidden)
        .chartXAxis(.hidden)
        // Animate threshold slide + bar recolor when the line changes.
        .animation(.easeInOut(duration: 0.25), value: line)
    }
}

File: wagerproof-android-native/app/src/main/java/com/wagerproof/app/features/props/detail/RecentPropBarChart.kt
package com.wagerproof.app.features.props.detail

import android.graphics.Paint
import android.graphics.Typeface
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.nativeCanvas
import androidx.compose.ui.graphics.toArgb
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.material3.Text
import com.wagerproof.core.design.tokens.AppColors
import com.wagerproof.core.models.MLBPlayerProps
import com.wagerproof.core.models.MLBPropChartBar

/**
 * Recent-form bar chart for a player prop — hand-drawn with Compose [Canvas]
 * (no external chart lib). As the selected line changes upstream, the dashed
 * threshold line slides (animated via [animateFloatAsState], easeInOut ~250ms)
 * and bars recolor green/red live against the animated threshold. Fixed
 * x-domain (oldest→newest); value annotations sit on each bar.
 * Port of iOS `RecentPropBarChart.swift`.
 */
@Composable
fun RecentPropBarChart(bars: List<MLBPropChartBar>, line: Double, modifier: Modifier = Modifier) {
    if (bars.isEmpty()) {
        Text(
            "No recent games",
            color = AppColors.appTextMuted,
            fontStyle = FontStyle.Italic,
            fontSize = 13.sp,
            modifier = modifier.fillMaxWidth(),
        )
        return
    }

    val animatedLine by animateFloatAsState(
        targetValue = line.toFloat(),
        animationSpec = tween(250),
        label = "propChartLine",
    )
    // Domain uses the target line so bars don't rescale mid-slide (RN parity).
    val maxVal = maxOf(line * 1.5, bars.maxOf { it.value }, line + 1, 1.0)

    val green = AppColors.appPrimary.toArgb()
    val loss = AppColors.appLoss
    val primary = AppColors.appPrimary
    val muted = AppColors.appTextMuted

    Column(modifier, verticalArrangement = Arrangement.spacedBy(4.dp)) {
        Canvas(Modifier.fillMaxWidth().height(168.dp)) {
            val n = bars.size
            val slot = size.width / n
            val barW = slot * 0.62f
            val topInset = 16.dp.toPx() // room for value annotations

            val valuePaint = Paint().apply {
                isAntiAlias = true
                textSize = 9.sp.toPx()
                typeface = Typeface.create(Typeface.DEFAULT, Typeface.BOLD)
                textAlign = Paint.Align.CENTER
            }

            bars.forEachIndexed { i, bar ->
                val cleared = bar.value > animatedLine
                val h = (bar.value / maxVal).toFloat() * (size.height - topInset)
                val cx = slot * i + slot / 2f
                val left = cx - barW / 2f
                val top = size.height - h
                val color = if (cleared) primary else loss.copy(alpha = 0.7f)
                drawRoundRect(
                    color = color,
                    topLeft = Offset(left, top),
                    size = Size(barW, h),
                    cornerRadius = CornerRadius(2.dp.toPx()),
                )
                valuePaint.color = if (cleared) green else loss.toArgb()
                drawContext.canvas.nativeCanvas.drawText(
                    MLBPlayerProps.formatBarValue(bar.value),
                    cx,
                    top - 4.dp.toPx(),
                    valuePaint,
                )
            }

            // Dashed threshold line.
            val ty = size.height - (animatedLine / maxVal).toFloat() * (size.height - topInset)
            drawLine(
                color = primary.copy(alpha = 0.85f),
                start = Offset(0f, ty),
                end = Offset(size.width, ty),
                strokeWidth = 1.2.dp.toPx(),
                pathEffect = PathEffect.dashPathEffect(floatArrayOf(4.dp.toPx(), 3.dp.toPx())),
            )
            val labelPaint = Paint().apply {
                isAntiAlias = true
                textSize = 9.sp.toPx()
                typeface = Typeface.create(Typeface.DEFAULT, Typeface.BOLD)
                textAlign = Paint.Align.RIGHT
                color = primary.toArgb()
            }
            drawContext.canvas.nativeCanvas.drawText(
                "Line ${MLBPlayerProps.formatBarValue(animatedLine.toDouble())}",
                size.width,
                ty - 3.dp.toPx(),
                labelPaint,
            )
        }

        // X labels (M/D), one per bar, oldest → newest.
        Row(Modifier.fillMaxWidth()) {
            bars.forEach { bar ->
                Text(
                    shortDate(bar.date) ?: "",
                    color = muted,
                    fontSize = 8.sp,
                    textAlign = TextAlign.Center,
                    modifier = Modifier.weight(1f),
                )
            }
        }
        Text(
            "Last ${bars.size} games · oldest left → most recent right",
            color = muted,
            fontSize = 10.sp,
            textAlign = TextAlign.Center,
            fontWeight = FontWeight.Normal,
            modifier = Modifier.fillMaxWidth(),
        )
    }
}

/** "2026-06-15" → "6/15". Mirrors RN `formatShortDate`. */
fun shortDate(iso: String?): String? {
    if (iso == null) return null
    val parts = iso.split("-")
    if (parts.size < 3) return null
    val month = parts[1].toIntOrNull() ?: return null
    val day = parts[2].take(2).toIntOrNull() ?: return null
    return "$month/$day"
}
